gulp-tasks/stageGitHub/cleanUpGoogleCloud.js

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Queries the Cloud Storage Bucket for currently deployed
+ * PRs, checks if there is a corresponding App Engine version (for full deploys)
+ * and then checks if the matching PR is still open on GitHub and otherwise
+ * removes the remains.
+ */
+
+const {Storage} = require('@google-cloud/storage');
+const {VersionsClient} = require('@google-cloud/appengine-admin');
+
+const {requestGitHubApi} = require('./lib/gitHubApi');
+
+/**
+ * The bucket where static files are deployed
+ */
+const BUCKET_NAME = 'static-dcc-staging';
+
+/**
+ * The project ID of the App Engine project
+ */
+const PROJECT_ID = 'dcc-staging';
+
+/** This folder contains the base version of the site and should
+ * never be deleted.
+ */
+const FALLBACK_DIRECTORY = 'main/';
+
+/**
+ * The version name of the default App Engine version, which should
+ * never be cleaned up (but also can't and would )
+ */
+const FALLBACK_VERSION = 'main';
+
+const storage = new Storage();
+const versionsClient = new VersionsClient();
+
+/**
+ * Reads all files from the root of a Google Cloud Storage bucket
+ * and returns the PR numbers of the deployed PRs.
+ * @returns {Promise<Array<number>>} [4124, 4125, 4126] for example
+ */
+async function getDeployedPullRequestsFromStorage() {
+  // eslint-disable-next-line no-unused-vars
+  const [files, nextQuery, apiResponse] = await storage
+    .bucket(BUCKET_NAME)
+    .getFiles({autoPaginate: false, delimiter: '/', prefix: ''});
+
+  const pullRequests = [];
+  // The prefixes (read: folder names) created by the Cloud Build job
+  // are in the form of pr-{number}, so they need to be cleaned up
+  for (const prefix of apiResponse.prefixes) {
+    if (prefix === FALLBACK_DIRECTORY) {
+      continue;
+    }
+    const prNumber = prefix.replace('pr-', '').replace('/', '');
+    pullRequests.push(parseInt(prNumber));
+  }
+
+  return pullRequests;
+}
+
+/**
+ * Fetches the open pull requests from GitHub and
+ * maps the API response to only get the pull request numbers.
+ * @returns {Promise<Array<number>>} [4124, 4125, 4126] for example
+ */
+async function getOpenPullRequestsFromGitHub() {
+  try {
+    const apiResponse = await requestGitHubApi(
+      'GET',
+      'pulls?state=open&per_page=100'
```

Comment:
I'd record that 100 in some top-level variable, to make it clear we only check last 100. I think 100 is sufficient, just to make it more future proof, in case someone wonders why not PRs are accoutned for

Reply:
Added a comment instead. As the job is intended to run periodically it will pick up overflown items with the next run, even if we wouldn't set per_page and stick to the default.